test/Dialect/FIRRTL/ref.mlir

```diff
@@ -133,3 +133,53 @@ firrtl.circuit "RefTypeVector" {
     %rb = firrtl.ref.resolve %12 : !firrtl.ref<uint<2>>
   }
 }
+
+// -----
+
+// https://github.com/llvm/circt/issues/3715
+firrtl.circuit "Issue3715" {
+  firrtl.module private @Test(in %p: !firrtl.uint<1>, out %x: !firrtl.ref<uint<2>>) {
+    firrtl.when %p : !firrtl.uint<1> {
+      %zero = firrtl.constant 1 : !firrtl.uint<2>
+      %w = firrtl.wire : !firrtl.uint<2>
+      %1 = firrtl.ref.send %w : !firrtl.uint<2>
+      firrtl.ref.define %x, %1 : !firrtl.ref<uint<2>>
+      firrtl.strictconnect %w, %zero : !firrtl.uint<2>
+    }
+  }
+  firrtl.module @Issue3715(in %p: !firrtl.uint<1>) {
+    %test_in, %test_x = firrtl.instance test @Test(in p: !firrtl.uint<1>, out x: !firrtl.ref<uint<2>>)
+    firrtl.strictconnect %test_in, %p : !firrtl.uint<1>
+    %x = firrtl.ref.resolve %test_x : !firrtl.ref<uint<2>>
+  }
+}
+
+// -----
+
+// Support using output port reference locally.
+// https://github.com/llvm/circt/issues/3713
+
+firrtl.circuit "UseRefsWithSinkFlow" {
+  firrtl.module private @InChild(in %p: !firrtl.ref<uint<1>>) {
+  }
+  firrtl.module private @OutChild(in %x: !firrtl.uint, out %y: !firrtl.uint, out %p: !firrtl.ref<uint>) {
+    %0 = firrtl.ref.send %x : !firrtl.uint
+    firrtl.ref.define %p, %0 : !firrtl.ref<uint>
+    %1 = firrtl.ref.resolve %p : !firrtl.ref<uint>
```

Comment:
Maybe we need to make sure `LowerXMR` can handle this.

Reply:
Good call.  It makes it through `firtool`, but directly running `LowerXMR` fails / crashes in a cast looks like, missed that.  The issue fixed is specifically about this being rejected by verifier as invalid entirely, which is what this test checks.

Or is there a change/action you're suggesting here?

Something for the TODO re:LowerXMR crashing on this?  LowerXMR will need reworking for handling ref.sub and u-turns (#4692) as well, (and aggregates-with-refs eventually), so think this goes on that list.